File: leetcode/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        this->preorder = &preorder;
        for (int i = 0; i < (int)inorder.size(); i++) {
            inorderIndex[inorder[i]] = i;
        }
        prePos = 0;

        return build(0, (int)inorder.size() - 1);
    }

private:
    vector<int>* preorder;
    unordered_map<int, int> inorderIndex;
    int prePos;

    TreeNode* build(int left, int right) {
        if (left > right) {
            return nullptr;
        }

        int rootVal = (*preorder)[prePos++];
        TreeNode* root = new TreeNode(rootVal);

        int mid = inorderIndex[rootVal];
        root->left = build(left, mid - 1);
        root->right = build(mid + 1, right);

        return root;
    }
};

File: leetcode/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inorder_index = {val: i for i, val in enumerate(inorder)}
        self.pre_pos = 0

        def build(left: int, right: int) -> Optional[TreeNode]:
            if left > right:
                return None

            root_val = preorder[self.pre_pos]
            self.pre_pos += 1
            root = TreeNode(root_val)

            mid = inorder_index[root_val]
            # everything left of mid in inorder is the left subtree, everything right is the right subtree
            root.left = build(left, mid - 1)
            root.right = build(mid + 1, right)

            return root

        return build(0, len(inorder) - 1)
